app/Libraries/TeamSpeak.php

```diff
@@ -275,26 +275,35 @@ public static function checkMemberMandatoryNotifications(TeamSpeak3_Node_Client
-            // if their nickname doesn't contain their name, or their grace period has ended
-            if (!$member->isPartiallyValidDisplayName($client['client_nickname']) || ($recentlyTold && !$hasGracePeriod)) {
-                self::pokeClient($client, trans('teamspeak.nickname.invalid.poke1'));
-                self::pokeClient($client, trans('teamspeak.nickname.invalid.poke2'));
-                self::kickClient($client, trans('teamspeak.nickname.invalid.kick'));
-                Cache::forget(self::CACHE_NICKNAME_PARTIALLY_CORRECT.$client['client_database_id']);
-                Cache::forget(self::CACHE_NICKNAME_PARTIALLY_CORRECT_GRACE.$client['client_database_id']);
-                throw new ClientKickedFromServerException;
-            } elseif (!$hasGracePeriod) {
-                // set grace period to allow for possible mistakes
-                self::pokeClient($client, trans('teamspeak.nickname.partiallyinvalid.poke1'));
-                self::pokeClient($client, trans('teamspeak.nickname.partiallyinvalid.poke2'));
-                self::messageClient($client, trans('teamspeak.nickname.partiallyinvalid.note', ['example' => $member->real_name.' - EGLL_N_TWR']));
-
-                $now = Carbon::now();
-                Cache::put(self::CACHE_NICKNAME_PARTIALLY_CORRECT.$client['client_database_id'], $now, 6 * 60);
-                Cache::put(self::CACHE_NICKNAME_PARTIALLY_CORRECT_GRACE.$client['client_database_id'], $now, 3 * 60);
+
+            // Check to see if their name is at least partially right
+            if ($member->isPartiallyValidDisplayName($client['client_nickname'])) {
+
+                // Check if they have been recently told but the grace has elapsed
+                if ($recentlyTold && !$hasGracePeriod) {
+                    self::pokeClient($client, trans('teamspeak.nickname.invalid.poke1'));
+                    self::pokeClient($client, trans('teamspeak.nickname.invalid.poke2'));
+                    self::kickClient($client, trans('teamspeak.nickname.invalid.kick'));
+                    Cache::forget(self::CACHE_NICKNAME_PARTIALLY_CORRECT.$client['client_database_id']);
+                    Cache::forget(self::CACHE_NICKNAME_PARTIALLY_CORRECT_GRACE.$client['client_database_id']);
+                    throw new ClientKickedFromServerException;
+                    return;
+                }
+
+                // Give them a grace period
+                if (!$hasGracePeriod) {
+                    self::pokeClient($client, trans('teamspeak.nickname.partiallyinvalid.poke1'));
+                    self::pokeClient($client, trans('teamspeak.nickname.partiallyinvalid.poke2'));
+                    self::messageClient($client, trans('teamspeak.nickname.partiallyinvalid.note', ['example' => $member->real_name.' - EGLL_N_TWR']));
+
+                    $now = Carbon::now();
+                    Cache::put(self::CACHE_NICKNAME_PARTIALLY_CORRECT.$client['client_database_id'], $now, 6 * 60);
```

Comment:
I'd be tempted to write some form of cache wrapper here (and elsewhere). I never like to see a direct call to the cache. Will leave this up to you though as I feel it might be a sizable job.

Reply:
This is probably outside of the scope - I am sure we can come back to this later (I do not have the will power to do more TeamSpeak messing right now!)